File: experiments/node/src/lib/writer-options.js
import { stringToUUID } from './binary.js';

export function dedupeEntryItems(items) {
  items.sort((a, b) => (a.offset < b.offset ? -1 : a.offset > b.offset ? 1 : 0));
  const [firstItem, ...remainingItems] = items;
  const deduped = [firstItem];
  let lastOffset = firstItem.offset;
  for (const item of remainingItems) {
    if (item.offset !== lastOffset) {
      deduped.push(item);
      lastOffset = item.offset;
    }
  }
  return deduped;
}

export function normalizeFileMode(value, defaultMode) {
  if (value === undefined || value === null) return defaultMode;
  if (!Number.isInteger(value) || value < 0 || value > 0o777) {
    throw new Error(`invalid journal file mode: ${value}`);
  }
  return value;
}

export function normalizeLivePublishEveryEntries(value) {
  if (value === undefined || value === null) return 1;
  if (!Number.isSafeInteger(value) || value < 0) {
    throw new Error(`invalid livePublishEveryEntries: ${value}`);
  }
  return value;
}

export function uuidOption(value, label) {
  if (value === undefined || value === null) return null;
  let out;
  if (typeof value === 'string') {
    const clean = value.trim().replaceAll('-', '');
    if (!/^[0-9a-fA-F]{32}$/.test(clean)) throw new Error(`${label} must be 16 bytes or 32 hex characters`);
    out = stringToUUID(clean);
  } else {
    out = Buffer.from(value);
  }
  if (out.length !== 16) throw new Error(`${label} must be 16 bytes or 32 hex characters`);
  return out;
}

File: experiments/python/journal/writer_options.py
def _current_time_ms():
    import time
    return int(time.time() * 1000)


def _dedupe_entry_items(items):
    deduped = [items[0]]
    for i in range(1, len(items)):
        if items[i]['offset'] != deduped[-1]['offset']:
            deduped.append(items[i])
    return deduped


def _uuid_option(value, fallback):
    if value is None:
        return fallback
    if isinstance(value, str):
        value = value.replace('-', '')
        return bytes.fromhex(value)
    if isinstance(value, bytearray):
        value = bytes(value)
    if not isinstance(value, bytes) or len(value) != 16:
        raise ValueError('uuid options must be 16 bytes or 32 hex characters')
    return value


def _normalize_live_publish_every_entries(value):
    if value is None:
        return 1
    entries = int(value)
    if entries < 0:
        raise ValueError(f'invalid live_publish_every_entries: {value}')
    return entries
